fast-list.js

```diff
@@ -72,42 +63,109 @@ function FastList(source) {
+      // the initialScrollTop to avoid
+      // triggering 'scroll' handler
+      schedule.attachDirect(
+        this.els.container,
+        'scroll',
+        this.handleScroll
+      );
+    }.bind(this));
+}
 
-    // We can't set initial scrollTop
-    // until the list has a height
-    if (source.initialScrollTop) {
-      this.els.container.scrollTop = source.initialScrollTop;
-      this.geometry.topPosition = this.els.container.scrollTop;
-    }
+FastList.prototype = {
+  PRERENDER_MULTIPLIER: 3.5,
 
-    // using a fragment container means we only
-    // endure one document mutation on inititalization
+  plugin: function(fn) {
+    fn(this);
+    return this;
+  },
+
+  /**
+   * The first render of the list aims
+   * to get content on the screen as
+   * fast as possible.
+   *
+   * Only the list-items in the viewport
+   * (critical) are rendered, and the
+   * populateItemDetail() stage is
+   * skipped to prevent expensive
+   * render paths being hit.
+   *
+   * @private
+   */
+  renderPhase1: function() {
+    debug('render phase 1');
     var fragment = document.createDocumentFragment();
+    var container = this.els.container;
+
+    // this.updateRenderingWindow(true);
+    this.updateContainerGeometry();
+
+    // Using a fragment container means we only
+    // endure one document mutation on inititalization
     this.updateSections({ fragment: fragment });
-    this.render({ fragment: fragment });
+
+    // Lightweight render into fragment
+    this.render({
+      fragment: fragment,
+      criticalOnly: true,
+      skipDetail: true
+    });
+
+    // If the list detached (created internally), attach it
+    if (!this.els.list.parentNode) container.appendChild(this.els.list);
+
+    // Inject the content fragment
     this.els.itemContainer.appendChild(fragment);
 
-    // Bind context early so that is can be detached later
-    this.handleScroll = this.handleScroll.bind(this);
+    // Make the container scrollable
+    container.style.overflowX = 'hidden';
+    container.style.overflowY = 'scroll';
+    this.updateListHeight();
 
-    // Bind scroll listeners after setting
-    // the initialScrollTop to avoid
-    // triggering 'scroll' handler
-    schedule.attachDirect(
-      this.els.container,
-      'scroll',
-      this.handleScroll
-    );
-  }.bind(this));
-}
+    // We can't set initial scrollTop
+    // until the list has a height
+    if (this.source.initialScrollTop) {
+      container.scrollTop = this.source.initialScrollTop;
+      this.geometry.topPosition = container.scrollTop;
+    }
+  },
 
-FastList.prototype = {
-  plugin: function(fn) {
-    fn(this);
-    return this;
+  /**
+   * The second render phase completes
+   * the initialization of the list.
+   *
+   * All list items will be rendered
+   * and the detail (eg. images) will
+   * be populated.
+   *
+   * @return {Promise}
+   * @private
+   */
+  renderPhase2: function() {
+    return new Promise(resolve => {
```

Comment:
almost forgot, but no fat arrow in the list library, we still have safari support :)